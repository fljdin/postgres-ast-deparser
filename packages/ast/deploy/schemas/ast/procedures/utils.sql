-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
