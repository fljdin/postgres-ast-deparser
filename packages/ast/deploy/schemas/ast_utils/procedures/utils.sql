-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
