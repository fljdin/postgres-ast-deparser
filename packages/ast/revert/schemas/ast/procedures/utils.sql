-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
