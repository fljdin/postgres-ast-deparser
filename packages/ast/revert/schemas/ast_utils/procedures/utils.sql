-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
