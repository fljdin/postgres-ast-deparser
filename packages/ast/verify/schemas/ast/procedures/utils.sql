-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
