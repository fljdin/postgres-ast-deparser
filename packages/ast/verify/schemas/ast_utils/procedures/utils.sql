-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
