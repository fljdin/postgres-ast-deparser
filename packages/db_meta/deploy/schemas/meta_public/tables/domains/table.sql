-- Deploy schemas/meta_public/tables/domains/table to pg

-- requires: schemas/meta_public/schema
-- requires: schemas/collections_public/tables/database/table 

BEGIN;

CREATE TABLE meta_public.domains (
    id uuid PRIMARY KEY DEFAULT uuid_generate_v4 (),
    database_id uuid NOT NULL REFERENCES collections_public.database (id) ON DELETE CASCADE,
    subdomain hostname,
    domain hostname,
    UNIQUE ( subdomain, domain )
);

CREATE INDEX domains_database_id_idx ON meta_public.domains ( database_id );

COMMIT;
