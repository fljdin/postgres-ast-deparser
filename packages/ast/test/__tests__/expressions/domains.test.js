import { getConnections } from '../../utils';
import { domains } from './__fixtures__/domains';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {}
});

// it('deparse', async () => {
//   const [
//     { deparse: result }
//   ] = await db.any(`select deparser.deparse( $1::jsonb );`, [domains[4]]);
//   expect(result).toMatchSnapshot();
// });

it('deparse', async () => {
  for (let i = 0; i < domains.length; i++) {
    const [
      { deparse: result }
    ] = await db.any(`select deparser.deparse( $1::jsonb );`, [domains[i]]);
    expect(result).toMatchSnapshot();
  }
});
