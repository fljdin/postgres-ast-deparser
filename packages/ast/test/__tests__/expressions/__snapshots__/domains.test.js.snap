// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deparse 1`] = `"CREATE DOMAIN domaindroptest AS int4"`;

exports[`deparse 2`] = `"COMMENT ON DOMAIN domaindoptest IS E'About to drop this..'"`;

exports[`deparse 3`] = `"CREATE DOMAIN dependenttypetest AS domaindroptest"`;

exports[`deparse 4`] = `"DROP DOMAIN domaindroptest CASCADE"`;

exports[`deparse 5`] = `"DROP DOMAIN domaindroptest CASCADE"`;

exports[`deparse 6`] = `"DROP DOMAIN domaindroptest CASCADE"`;

exports[`deparse 7`] = `"CREATE DOMAIN domainvarchar AS varchar(5)"`;

exports[`deparse 8`] = `"CREATE DOMAIN domainnumeric AS numeric(8, 2)"`;

exports[`deparse 9`] = `"CREATE DOMAIN domainint4 AS int4"`;

exports[`deparse 10`] = `"CREATE DOMAIN domaintext AS text"`;

exports[`deparse 11`] = `
"CREATE TABLE basictest (
 testint4 domainint4,
testtext domaintext,
testvarchar domainvarchar,
testnumeric domainnumeric 
)"
`;

exports[`deparse 12`] = `"SELECT COALESCE(4::domainint4, 7) IN ( int4 ) AS t"`;

exports[`deparse 13`] = `"SELECT COALESCE(4::domainint4, 7) IN ( domainint4 ) AS f"`;

exports[`deparse 14`] = `"SELECT COALESCE(4::domainint4, 7::domainint4) IN ( domainint4 ) AS t"`;

exports[`deparse 15`] = `"DROP TABLE basictest CASCADE"`;

exports[`deparse 16`] = `"DROP DOMAIN domainvarchar CASCADE"`;

exports[`deparse 17`] = `"DROP DOMAIN domainnumeric CASCADE"`;

exports[`deparse 18`] = `"DROP DOMAIN domainint4 CASCADE"`;

exports[`deparse 19`] = `"DROP DOMAIN domaintext CASCADE"`;

exports[`deparse 20`] = `"CREATE DOMAIN domainint4arr AS int4[]"`;

exports[`deparse 21`] = `"CREATE DOMAIN domainchar4arr AS varchar(4)[]"`;

exports[`deparse 22`] = `"DROP DOMAIN domainint4arr CASCADE"`;

exports[`deparse 23`] = `"DROP DOMAIN domainchar4arr CASCADE"`;

exports[`deparse 24`] = `"CREATE DOMAIN dia AS int[]"`;

exports[`deparse 25`] = `"DROP DOMAIN dia CASCADE"`;

exports[`deparse 26`] = `"CREATE DOMAIN dnotnull AS varchar(15) NOT NULL"`;

exports[`deparse 27`] = `"CREATE DOMAIN dnull AS varchar(15)"`;

exports[`deparse 28`] = `"CREATE DOMAIN dcheck AS varchar(15) NOT NULL, CHECK ((value = 'a' OR value = 'c' OR value = 'd'))"`;

exports[`deparse 29`] = `
"CREATE TABLE nulltest (
 col1 dnotnull,
col2 dnotnull NULL,
col3 dnull NOT NULL,
col4 dnull,
col5 dcheck CHECK (col5 IN ( 'c', 'd' )) 
)"
`;

exports[`deparse 30`] = `"INSERT INTO nulltest DEFAULT VALUES"`;

exports[`deparse 31`] = `"INSERT INTO nulltest VALUES ('a', 'b', 'c', 'd', 'c')"`;

exports[`deparse 32`] = `"INSERT INTO nulltest VALUES ('a', 'b', 'c', 'd', NULL)"`;

exports[`deparse 33`] = `"INSERT INTO nulltest VALUES ('a', 'b', 'c', 'd', 'a')"`;

exports[`deparse 34`] = `"INSERT INTO nulltest VALUES (NULL, 'b', 'c', 'd', 'd')"`;

exports[`deparse 35`] = `"INSERT INTO nulltest VALUES ('a', NULL, 'c', 'd', 'c')"`;

exports[`deparse 36`] = `"INSERT INTO nulltest VALUES ('a', 'b', NULL, 'd', 'c')"`;

exports[`deparse 37`] = `"INSERT INTO nulltest VALUES ('a', 'b', 'c', NULL, 'd')"`;

exports[`deparse 38`] = `"SELECT * nulltest"`;

exports[`deparse 39`] = `"SELECT '1'::dnotnull"`;

exports[`deparse 40`] = `"SELECT NULL::dnotnull"`;

exports[`deparse 41`] = `"SELECT NULL::dnull::dnotnull"`;

exports[`deparse 42`] = `"SELECT col4::dnotnull nulltest"`;

exports[`deparse 43`] = `"DROP TABLE nulltest CASCADE"`;

exports[`deparse 44`] = `"DROP DOMAIN dnotnull CASCADE"`;

exports[`deparse 45`] = `"DROP DOMAIN dnull CASCADE"`;

exports[`deparse 46`] = `"DROP DOMAIN dcheck CASCADE"`;

exports[`deparse 47`] = `"CREATE DOMAIN ddef1 AS int4 DEFAULT (3)"`;

exports[`deparse 48`] = `"CREATE DOMAIN ddef2 AS oid DEFAULT ('12')"`;

exports[`deparse 49`] = `"CREATE DOMAIN ddef3 AS text DEFAULT (5)"`;

exports[`deparse 50`] = `"CREATE SEQUENCE ddef4_seq"`;

exports[`deparse 51`] = `"CREATE DOMAIN ddef4 AS int4 DEFAULT (nextval ('ddef4_seq'))"`;

exports[`deparse 52`] = `"CREATE DOMAIN ddef5 AS numeric(8, 2) NOT NULL, DEFAULT ('12.12')"`;

exports[`deparse 53`] = `
"CREATE TABLE defaulttest (
 col1 ddef1,
col2 ddef2,
col3 ddef3,
col4 ddef4 PRIMARY KEY,
col5 ddef1 NOT NULL DEFAULT (NULL),
col6 ddef2 DEFAULT ('88'),
col7 ddef4 DEFAULT (8000),
col8 ddef5 
)"
`;

exports[`deparse 54`] = `"INSERT INTO defaulttest (col4) VALUES (0)"`;

exports[`deparse 55`] = `"ALTER TABLE defaulttest ALTER COLUMN col5 DROP DEFAULT"`;

exports[`deparse 56`] = `"INSERT INTO defaulttest DEFAULT VALUES"`;

exports[`deparse 57`] = `"ALTER TABLE defaulttest ALTER COLUMN col5 SET DEFAULT NULL"`;

exports[`deparse 58`] = `"INSERT INTO defaulttest (col4) VALUES (0)"`;

exports[`deparse 59`] = `"ALTER TABLE defaulttest ALTER COLUMN col5 DROP DEFAULT"`;

exports[`deparse 60`] = `"INSERT INTO defaulttest DEFAULT VALUES"`;

exports[`deparse 61`] = `"INSERT INTO defaulttest DEFAULT VALUES"`;

exports[`deparse 62`] = `"SELECT * defaulttest"`;

exports[`deparse 63`] = `"DROP TABLE defaulttest CASCADE"`;

exports[`deparse 64`] = `"CREATE DOMAIN dnotnulltest AS int"`;

exports[`deparse 65`] = `
"CREATE TABLE domnotnull (
 col1 dnotnulltest,
col2 dnotnulltest 
)"
`;

exports[`deparse 66`] = `"INSERT INTO domnotnull DEFAULT VALUES"`;

exports[`deparse 67`] = `"ALTER DOMAIN dnotnulltest CASCADE"`;

exports[`deparse 68`] = `"UPDATE domnotnull SET col1"`;

exports[`deparse 69`] = `"ALTER DOMAIN dnotnulltest CASCADE"`;

exports[`deparse 70`] = `"UPDATE domnotnull SET col2"`;

exports[`deparse 71`] = `"ALTER DOMAIN dnotnulltest CASCADE"`;

exports[`deparse 72`] = `"UPDATE domnotnull SET col1"`;

exports[`deparse 73`] = `"ALTER DOMAIN dnotnulltest CASCADE"`;

exports[`deparse 74`] = `"UPDATE domnotnull SET col1"`;

exports[`deparse 75`] = `"DROP DOMAIN dnotnulltest CASCADE"`;

exports[`deparse 76`] = `
"CREATE TABLE domdeftest (
 col1 ddef1 
)"
`;

exports[`deparse 77`] = `"INSERT INTO domdeftest DEFAULT VALUES"`;

exports[`deparse 78`] = `"SELECT * domdeftest"`;

exports[`deparse 79`] = `"ALTER DOMAIN ddef1 CASCADE"`;

exports[`deparse 80`] = `"INSERT INTO domdeftest DEFAULT VALUES"`;

exports[`deparse 81`] = `"SELECT * domdeftest"`;

exports[`deparse 82`] = `"ALTER DOMAIN ddef1 CASCADE"`;

exports[`deparse 83`] = `"INSERT INTO domdeftest DEFAULT VALUES"`;

exports[`deparse 84`] = `"SELECT * domdeftest"`;

exports[`deparse 85`] = `"DROP TABLE domdeftest CASCADE"`;

exports[`deparse 86`] = `"CREATE DOMAIN con AS int"`;

exports[`deparse 87`] = `
"CREATE TABLE domcontest (
 col1 con 
)"
`;

exports[`deparse 88`] = `"INSERT INTO domcontest VALUES (1)"`;

exports[`deparse 89`] = `"INSERT INTO domcontest VALUES (2)"`;

exports[`deparse 90`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 91`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 92`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 93`] = `"INSERT INTO domcontest VALUES (-5)"`;

exports[`deparse 94`] = `"INSERT INTO domcontest VALUES (42)"`;

exports[`deparse 95`] = `"INSERT INTO domcontest VALUES (5)"`;

exports[`deparse 96`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 97`] = `"INSERT INTO domcontest VALUES (-5)"`;

exports[`deparse 98`] = `"INSERT INTO domcontest VALUES (42)"`;

exports[`deparse 99`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 100`] = `"ALTER DOMAIN con CASCADE"`;

exports[`deparse 101`] = `"CREATE DOMAIN things AS int"`;

exports[`deparse 102`] = `
"CREATE TABLE thethings (
 stuff things 
)"
`;

exports[`deparse 103`] = `"INSERT INTO thethings (stuff) VALUES (55)"`;

exports[`deparse 104`] = `"ALTER DOMAIN things CASCADE"`;

exports[`deparse 105`] = `"ALTER DOMAIN things CASCADE"`;

exports[`deparse 106`] = `"ALTER DOMAIN things CASCADE"`;

exports[`deparse 107`] = `"UPDATE thethings SET stuff"`;

exports[`deparse 108`] = `"ALTER DOMAIN things CASCADE"`;

exports[`deparse 109`] = `
"CREATE TABLE domtab (
 col1 int 
)"
`;

exports[`deparse 110`] = `"CREATE DOMAIN dom AS int"`;

exports[`deparse 111`] = `"CREATE VIEW domview AS SELECT col1::dom domtab"`;

exports[`deparse 112`] = `"INSERT INTO domtab (col1) VALUES (NULL)"`;

exports[`deparse 113`] = `"INSERT INTO domtab (col1) VALUES (5)"`;

exports[`deparse 114`] = `"SELECT * domview"`;

exports[`deparse 115`] = `"ALTER DOMAIN dom CASCADE"`;

exports[`deparse 116`] = `"SELECT * domview"`;

exports[`deparse 117`] = `"ALTER DOMAIN dom CASCADE"`;

exports[`deparse 118`] = `"SELECT * domview"`;

exports[`deparse 119`] = `"ALTER DOMAIN dom CASCADE"`;

exports[`deparse 120`] = `"SELECT * domview"`;

exports[`deparse 121`] = `"ALTER DOMAIN dom CASCADE"`;

exports[`deparse 122`] = `"SELECT * domview"`;

exports[`deparse 123`] = `"DROP DOMAIN ddef1 CASCADE"`;

exports[`deparse 124`] = `"DROP DOMAIN ddef2 CASCADE"`;

exports[`deparse 125`] = `"DROP DOMAIN ddef3 CASCADE"`;

exports[`deparse 126`] = `"DROP DOMAIN ddef4 CASCADE"`;

exports[`deparse 127`] = `"DROP DOMAIN ddef5 CASCADE"`;

exports[`deparse 128`] = `"DROP SEQUENCE ddef4_seq CASCADE"`;

exports[`deparse 129`] = `"CREATE DOMAIN vchar4 AS varchar(4)"`;

exports[`deparse 130`] = `"CREATE DOMAIN dinter AS vchar4 CHECK (pg_catalog.substring (value, 1, 1) = 'x')"`;

exports[`deparse 131`] = `"CREATE DOMAIN dtop AS dinter CHECK (pg_catalog.substring (value, 2, 1) = '1')"`;

exports[`deparse 132`] = `"SELECT 'x123'::dtop"`;

exports[`deparse 133`] = `"SELECT 'x1234'::dtop"`;

exports[`deparse 134`] = `"SELECT 'y1234'::dtop"`;

exports[`deparse 135`] = `"SELECT 'y123'::dtop"`;

exports[`deparse 136`] = `"SELECT 'yz23'::dtop"`;

exports[`deparse 137`] = `"SELECT 'xz23'::dtop"`;

exports[`deparse 138`] = `
"CREATE dtest (
 f1 dtop 
)"
`;

exports[`deparse 139`] = `"INSERT INTO dtest VALUES ('x123')"`;

exports[`deparse 140`] = `"INSERT INTO dtest VALUES ('x1234')"`;

exports[`deparse 141`] = `"INSERT INTO dtest VALUES ('y1234')"`;

exports[`deparse 142`] = `"INSERT INTO dtest VALUES ('y123')"`;

exports[`deparse 143`] = `"INSERT INTO dtest VALUES ('yz23')"`;

exports[`deparse 144`] = `"INSERT INTO dtest VALUES ('xz23')"`;

exports[`deparse 145`] = `"DROP TABLE dtest CASCADE"`;

exports[`deparse 146`] = `"DROP DOMAIN vchar4 CASCADE"`;

exports[`deparse 147`] = `"CREATE DOMAIN str_domain AS text NOT NULL"`;

exports[`deparse 148`] = `
"CREATE TABLE domain_test (
 a int,
b int 
)"
`;

exports[`deparse 149`] = `"INSERT INTO domain_test VALUES (1, 2)"`;

exports[`deparse 150`] = `"INSERT INTO domain_test VALUES (1, 2)"`;

exports[`deparse 151`] = `"ALTER TABLE domain_test ADD COLUMN c str_domain"`;

exports[`deparse 152`] = `"CREATE DOMAIN str_domain2 AS text CHECK (value <> 'foo'), DEFAULT ('foo')"`;

exports[`deparse 153`] = `"ALTER TABLE domain_test ADD COLUMN d str_domain2"`;

exports[`deparse 154`] = `"CREATE DOMAIN pos_int AS int4 CHECK (value > 0), NOT NULL"`;

exports[`deparse 155`] = `"s1 (10)"`;

exports[`deparse 156`] = `"s1 (0)"`;

exports[`deparse 157`] = `"s1 (NULL)"`;

exports[`deparse 158`] = `
"CREATE FUNCTION doubledecrement ( p1 pos_int ) RETURNS pos_int AS $LQLCODEZ$ 
 
declare v pos_int;
begin
    return p1;
end 
 $LQLCODEZ$ LANGUAGE plpgsql"
`;

exports[`deparse 159`] = `"SELECT doubledecrement (3)"`;

exports[`deparse 160`] = `
"CREATE OR REPLACE FUNCTION doubledecrement ( p1 pos_int ) RETURNS pos_int AS $LQLCODEZ$ 
 
declare v pos_int := 0;
begin
    return p1;
end 
 $LQLCODEZ$ LANGUAGE plpgsql"
`;

exports[`deparse 161`] = `"SELECT doubledecrement (3)"`;

exports[`deparse 162`] = `
"CREATE OR REPLACE FUNCTION doubledecrement ( p1 pos_int ) RETURNS pos_int AS $LQLCODEZ$ 
 
declare v pos_int := 1;
begin
    v := p1 - 1;
    return v - 1;
end 
 $LQLCODEZ$ LANGUAGE plpgsql"
`;

exports[`deparse 163`] = `"SELECT doubledecrement (NULL)"`;

exports[`deparse 164`] = `"SELECT doubledecrement (0)"`;

exports[`deparse 165`] = `"SELECT doubledecrement (1)"`;

exports[`deparse 166`] = `"SELECT doubledecrement (2)"`;

exports[`deparse 167`] = `"SELECT doubledecrement (3)"`;

exports[`deparse 168`] = `"CREATE DOMAIN posint AS int4"`;

exports[`deparse 169`] = `"CREATE TYPE ddtest1 AS (f1 posint)"`;

exports[`deparse 170`] = `
"CREATE TABLE ddtest2 (
 f1 ddtest1 
)"
`;

exports[`deparse 171`] = `"INSERT INTO ddtest2 VALUES (ROW(-1))"`;

exports[`deparse 172`] = `"ALTER DOMAIN posint CASCADE"`;

exports[`deparse 173`] = `"DROP TABLE ddtest2 CASCADE"`;

exports[`deparse 174`] = `
"CREATE TABLE ddtest2 (
 f1 ddtest1[] 
)"
`;

exports[`deparse 175`] = `"INSERT INTO ddtest2 VALUES ('{(-1)}')"`;

exports[`deparse 176`] = `"ALTER DOMAIN posint CASCADE"`;

exports[`deparse 177`] = `"ALTER DOMAIN posint CASCADE"`;

exports[`deparse 178`] = `"CREATE DOMAIN posint2 AS posint CHECK (value % 2 = 0)"`;

exports[`deparse 179`] = `"ALTER DOMAIN posint CASCADE"`;

exports[`deparse 180`] = `"ALTER DOMAIN posint CASCADE"`;

exports[`deparse 181`] = `"DROP TABLE ddtest2 CASCADE"`;

exports[`deparse 182`] = `"DROP TYPE ddtest1 CASCADE"`;

exports[`deparse 183`] = `"DROP DOMAIN posint CASCADE"`;

exports[`deparse 184`] = `"CREATE DOMAIN mynums AS numeric(4, 2)[]"`;

exports[`deparse 185`] = `"CREATE DOMAIN mynums2 AS mynums"`;

exports[`deparse 186`] = `"CREATE DOMAIN orderedpair AS int[] CHECK ([1] < [2])"`;

exports[`deparse 187`] = `"CREATE DOMAIN di AS int"`;

exports[`deparse 188`] = `"ALTER DOMAIN di CASCADE"`;

exports[`deparse 189`] = `"CREATE DOMAIN inotnull AS int CHECK (sql_is_distinct_from (value, NULL))"`;

exports[`deparse 190`] = `"DROP DOMAIN inotnull CASCADE"`;

exports[`deparse 191`] = `"CREATE DOMAIN testdomain1 AS int"`;

exports[`deparse 192`] = `""`;

exports[`deparse 193`] = `""`;

exports[`deparse 194`] = `"DROP DOMAIN testdomain3 CASCADE"`;

exports[`deparse 195`] = `"CREATE DOMAIN testdomain1 AS int CONSTRAINT unsigned CHECK (value > 0)"`;

exports[`deparse 196`] = `""`;

exports[`deparse 197`] = `"ALTER DOMAIN testdomain1 CASCADE"`;

exports[`deparse 198`] = `"DROP DOMAIN testdomain1 CASCADE"`;
